portal/src/util/error.ts

```diff
@@ -0,0 +1,175 @@
+import { ApolloError } from "@apollo/client";
+import { GraphQLError } from "graphql";
+import { nonNullable } from "./types";
+import { Values } from "@oursky/react-messageformat";
+
+// union type of different kind of violation
+export type Violation = RequiredViolation;
+
+interface RequiredViolation {
+  kind: "required";
+  location: string;
+  missingField: string[];
+}
+
+// list of violation kind recognized
+const violationKinds = ["required"];
+type ViolationKind = Violation["kind"];
+function isViolationKind(value?: string): value is ViolationKind {
+  return value != null && violationKinds.includes(value);
+}
+
+// expected data shape of error extension from backend
+interface ErrorCauseDetails {
+  actual?: string[];
+  expected?: string[];
+  missing?: string[];
+}
+
+interface ErrorCause {
+  details?: ErrorCauseDetails;
+  location?: string;
+  kind?: string;
+}
+
+interface ErrorExtensionInfo {
+  causes?: ErrorCause[];
+}
+
+interface GraphQLErrorExtensions {
+  errorName: string;
+  info: ErrorExtensionInfo;
+  reason: string;
+}
+
+type ViolationSelector = (violation: Violation) => boolean;
+type ViolationSelectors<Key extends string> = Record<Key, ViolationSelector>;
```

Comment:
It is better to split this file into two.

The first file should be named `error.ts`, for parsing error into recognized shapes.

The second one should be named `validation.ts`, the validation error related type definitions and functions should go here.